### PR TITLE
Add our own DB exception abstraction

### PR DESCRIPTION
**Pull request by ChristophWurst**
Right now our API exports the Doctrine/dbal exception. As we've seen
with the dbal 3 upgrade, the leakage of 3rdparty types is problematic as
a dependency update means lots of work in apps, due to the direct
dependency of what Nextcloud ships. This breaks this dependency so that
apps only need to depend on our public API. That API can then be vendor
(db lib) agnostic and we can work around future deprecations/removals in
dbal more easily.

Right now the type of exception thrown is transported as "reason". For
the more popular types of errors we can extend the new exception class
and allow apps to catch specific errors only. Right now they have to
catch-check-rethrow. This is not ideal, but better than the dependnecy
on dbal.

Follow-up to #24948 

* [x] add to #23210

### CODE REVIEW

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

File: lib/private/DB/ConnectionAdapter.php [hunk 1]
```diff
@@ -25,12 +25,15 @@
 
 namespace OC\DB;
 
+use Doctrine\DBAL\Exception;
```
Comment (nickvergessen) [inline lib/private/DB/ConnectionAdapter.php:28]: something complains on the imports

File: lib/private/DB/Migrator.php [hunk 2]
```diff
@@ -82,6 +82,8 @@ public function __construct(\Doctrine\DBAL\Connection $connection,
 
 	/**
 	 * @param \Doctrine\DBAL\Schema\Schema $targetSchema
+	 *
+	 * @throws Exception
```
Comment (nickvergessen) [inline lib/private/DB/Migrator.php:86]: missing import now that its in OCP?

**Review by nickvergessen — CHANGES_REQUESTED** [review]
👍🏼 apart of the comment

File: lib/private/DB/ConnectionAdapter.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/private/DB/ConnectionAdapter.php:28]: there was one unused below

File: lib/private/DB/Migrator.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/private/DB/Migrator.php:86]: no, that seems fine

Comment (nickvergessen) [reply lib/private/DB/Migrator.php:86]: So `Doctrine\DBAL\Exception` is fine, not our exception? But looking this directly works with `\Doctrine\DBAL\Connection` I guess so.

**Review by nickvergessen — APPROVED** [review]
This will hit like a truck 🙈

File: lib/private/DB/Migrator.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/private/DB/Migrator.php:86]: Basically I kept it to Doctrine errors internally and only *convert* when they bubble up to the public API.

**Comment by ChristophWurst** [comment]
> This will hit like a truck see_no_evil

Oh yes. But to be fair not only because of this PR. It will already hit hard since the original doctrine exception vanished. So it's just a matter of which new exception to listen for. And some of the now thrown exceptions were thrown before too. They were just not documented in our abstraction, but very present in the Doctrine signature.

**Comment by nickvergessen** [comment]
Well all the `UniqueConstraintViolation` and so on still exist and worked (until this PR) Only the catch all didn't

**Comment by ChristophWurst** [comment]
Right. I hadn't factored that in. But then these special exceptions were never documented to be thrown … :stuck_out_tongue:

**Comment by MorrisJobke** [comment]
Psalm says no

**Comment by ChristophWurst** [comment]
> Psalm says no

Look again. It's unrelated.

File: lib/private/DB/Connection.php [hunk 3]
```diff
@@ -364,7 +377,9 @@ public function setValues($table, array $keys, array $values, array $updatePreco
 	 * Create an exclusive read+write lock on a table
 	 *
 	 * @param string $tableName
-	 * @throws \BadMethodCallException When trying to acquire a second lock
+	 *
+	 * @throws \BadMethodCallException When trying to acquire a second lock*@throws Exception
```
Comment (MorrisJobke) [inline lib/private/DB/Connection.php:381]:
```suggestion
	 * @throws \BadMethodCallException When trying to acquire a second lock
```

Comment (ChristophWurst) [reply lib/private/DB/Connection.php:381]: I blame phpstorm

**Review by MorrisJobke — APPROVED** [review]

File: lib/public/DB/Exception.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	 * @since 21.0.0
+	 */
+	public const REASON_CONNECTION_LOST = 1;
+
+	/**
+	 * A database constraint was violated
+	 *
+	 * @since 21.0.0
+	 */
+	public const REASON_CONSTRAINT_VIOLATION = 2;
```
Comment (rullzer) [inline lib/public/DB/Exception.php:55]: so the plan is first to have this generic and then maybe at a later stage have constraint violation exceptions? Or just document it properly and have the callers check the reason?

Comment (ChristophWurst) [reply lib/public/DB/Exception.php:55]: Yes and yes. The generic way is that the just catch this exception and optonally compare the reason to a constant. And later we can have specific subclasses that can be caught

**Review by rullzer — APPROVED** [review]

**Pull request merged**